source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <iostream>
+#include <memory>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
+
+namespace quic {
+
+class QuicMemSliceImpl {
+public:
+  // Constructs an empty QuicMemSliceImpl.
+  QuicMemSliceImpl() = default;
+
+  // Constructs a QuicMemSliceImpl by let |allocator| allocate a data buffer of
+  // |length|.
+  QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length)
+      : length_(length), slice_(new Envoy::Buffer::OwnedImpl()) {
+    auto fragment = new Envoy::Buffer::BufferFragmentImpl(
+        allocator->New(length), length,
+        [allocator](const void* data, size_t, const Envoy::Buffer::BufferFragmentImpl* fragment) {
+          allocator->Delete(const_cast<char*>(static_cast<const char*>(data)));
+          delete fragment;
+        });
+    slice_->addBufferFragment(*fragment);
+  }
+
+  // Constructs a QuicMemSliceImpl from a Buffer::Instance whose getRawSlices()
+  // returns only 1 slice.
+  QuicMemSliceImpl(std::shared_ptr<Envoy::Buffer::Instance> slice)
```

Comment:
take this as a const shared_ptr& to avoid having the infrastructure bump and drop the reference count.

Reply:
Not need for shared_ptr now. Changed to pass in a reference of Instance.